Add Google Maps for Work signing support

XanderStrike opened the pull request:
The Google Maps API for businesses requires signing requests according
to this specification [1]. This change allows signing such requests.

In addition we've added one error handling case that provides more
information when a 403 response is returned as the previous response was
misleading.

1. https://developers.google.com/maps/documentation/business/webservices/

panthomakos commented on README.markdown:
```diff
@@ -78,6 +78,7 @@ Next, add the following to your application.rb file, or before you perform a coo
 
     Timezone::Configure.begin do |c|
       c.google_api_key = 'your_google_api_key_goes_here'
+      c.google_client = 'your_google_client_id' # only if using 'Google for Work'
```
Using `google_client_id` would be preferable since it more clearly matches the documentation and helps indicate that it's an identifier instead of just an object.

panthomakos commented on test/google_lookup_test.rb:
```diff
@@ -39,6 +39,33 @@ def test_google_request_denied_read_lat_lon_coordinates
+    assert_equal '/maps/api/timezone/json?location=123%2C123', result[0]
+    assert result[1].start_with?('timestamp=')
+    assert_equal 'key=123abc', result[2]
+    assert_equal 3, result.count
```
Please use timecop here to check the timestamp, or just check for the `key` parameter and ignore everything else:

```
result = URI(lookup.send(:url, '123', '123')).query.split('&')
assert result.include?('key=123abc')
```

Ideally `url` won't get tested at all since it's a private method, but I can work on improving the `HTTPTestClient` to deal with that.

panthomakos commented on test/google_lookup_test.rb:
```diff
@@ -39,6 +39,33 @@ def test_google_request_denied_read_lat_lon_coordinates
+    assert_equal '/maps/api/timezone/json?location=123%2C123', result[0]
+    assert result[1].start_with?('timestamp=')
+    assert_equal 'client=123%26asdf', result[2]
+    assert result[3].start_with?('signature=')
```
Would be nice to have a signature validation here. Check out using `timecop` to validate that you are properly computing the signature.

panthomakos commented on lib/timezone/lookup/google.rb:
```diff
@@ -16,6 +19,10 @@ def initialize(config)
         response = client.get(url(lat,lng))
 
+        if response.code == '403'
+          raise(Timezone::Error::Google, response.body)
```
This makes me nervous. I don't like including an entire HTTP response body as an error message. I would be happy to look at a PR that includes a logging/debugging configuration option:

```
def lookup(lat, lng)
  if response.body == '403'
    config.logger.debug("Timezone::Error::Google : #{response.body}") if config.debug?
    raise(Timezone::Error::Google, '403 Forbidden')
  end
end
```

panthomakos commented:
Thanks for contributing!

XanderStrike commented:
Great feedback! I renamed `google_client` to `google_client_id`, changed the 403 error message to not include the response body (I agree that was a bit silly), and used Timecop in the URL tests.

Thanks so much for making this project, and reviewing my PR!

XanderStrike commented:
Oops! The build was failing on older versions of Ruby. Turns out the test key (`123abc`) isn't a valid base64 string and between 2.0.0 and 2.1.0 the handling of that changed. I've changed it to a real base64 number and we're in the green.

panthomakos commented:
Very nice! Thanks again for the contribution.
